gossip/src/push_active_set.rs

```diff
@@ -0,0 +1,237 @@
+use {
+    crate::weighted_shuffle::WeightedShuffle,
+    indexmap::IndexMap,
+    rand::Rng,
+    solana_bloom::bloom::{AtomicBloom, Bloom},
+    solana_sdk::{native_token::LAMPORTS_PER_SOL, pubkey::Pubkey},
+    std::collections::HashMap,
+};
+
+const NUM_PUSH_ACTIVE_SET_ENTRIES: usize = 36;
+
+pub(crate) struct PushActiveSet([PushActiveSetEntry; NUM_PUSH_ACTIVE_SET_ENTRIES]);
+
+// Keys are gossip nodes to push messages to.
+// Values are which origins the node has pruned.
+#[derive(Default)]
+struct PushActiveSetEntry(IndexMap</*node:*/ Pubkey, /*origins:*/ AtomicBloom<Pubkey>>);
+
+impl PushActiveSet {
+    #[cfg(debug_assertions)]
+    const MIN_NUM_BLOOM_ITEMS: usize = 512;
+    #[cfg(not(debug_assertions))]
+    const MIN_NUM_BLOOM_ITEMS: usize = crate::cluster_info::CRDS_UNIQUE_PUBKEY_CAPACITY;
+
+    pub(crate) fn get<'a>(
+        &'a self,
+        pubkey: &Pubkey,    // This node.
+        origin: &'a Pubkey, // CRDS value owner.
+        // If true forces gossip push even if the node has pruned the origin.
+        should_force_push: impl FnMut(&Pubkey) -> bool + 'a,
+        stakes: &HashMap<Pubkey, u64>,
+    ) -> impl Iterator<Item = &Pubkey> + 'a {
+        let stake = stakes.get(pubkey).min(stakes.get(origin));
+        self.get_entry(stake).get(origin, should_force_push)
+    }
+
+    pub(crate) fn prune(
+        &self,
+        pubkey: &Pubkey,    // This node.
+        node: &Pubkey,      // Gossip node.
+        origins: &[Pubkey], // CRDS value owners.
+        stakes: &HashMap<Pubkey, u64>,
+    ) {
+        let stake = stakes.get(pubkey);
+        for origin in origins {
+            if origin == pubkey {
+                continue;
+            }
+            let stake = stake.min(stakes.get(origin));
+            self.get_entry(stake).prune(node, origin)
+        }
+    }
+
+    pub(crate) fn rotate<R: Rng>(
+        &mut self,
+        rng: &mut R,
+        size: usize, // Number of nodes to retain in each active-set entry.
+        cluster_size: usize,
+        nodes: &[Pubkey],
+        stakes: &HashMap<Pubkey, u64>,
+    ) {
+        let num_bloom_filter_items = cluster_size.max(Self::MIN_NUM_BLOOM_ITEMS);
+        let weights: Vec<_> = nodes
+            .iter()
+            .map(|node| {
+                let stake = stakes.get(node).copied().unwrap_or_default();
+                let stake = stake / LAMPORTS_PER_SOL;
+                (u64::BITS - stake.leading_zeros()) as u64
+            })
+            .collect();
+        for (k, entry) in self.0.iter_mut().enumerate() {
+            let weights: Vec<u64> = weights
+                .iter()
+                .map(|&weight| weight.min(k as u64).saturating_add(1).saturating_pow(2))
```

Comment:
not sure i understand the math here.  a comment might help

Reply:
Added comments:
```
bucket <- get_stake_bucket(min stake of {
   this node, crds value owner and gossip peer
})
weight <- (bucket + 1)^2
```
`min stake of {...}` is a proxy for how much we care about this link, and tries to mirror similar logic on the receiving end when pruning incoming links:
https://github.com/solana-labs/solana/blob/81394cf92/gossip/src/received_cache.rs#L100-L105

It is pretty heuristic but we need something here and we can re-evaluate and adjust.

Reply:
thanks that helps!  mind adding this part to the code comment?

> a proxy for how much we care about this link, and tries to mirror similar logic on the receiving end when pruning incoming

Reply:
done.